src/systems/environment_preload/VisualizationTool.cc

```diff
@@ -0,0 +1,211 @@
+    const UpdateInfo &_info,
+    const EntityComponentManager& _ecm,
+    const std::shared_ptr<components::EnvironmentalData> _data,
+    double _xSamples, double _ySamples, double _zSamples)
+{
+
+  auto now = std::chrono::steady_clock::now();
+  std::chrono::duration<double> dt(now - this->lastTick);
+
+  if (this->resample)
+  {
+    this->CreatePointCloudTopics(_data);
+    this->ResizeCloud(_data, _ecm, _xSamples, _ySamples, _zSamples);
+    this->resample = false;
+    this->lastTick = now;
+  }
+
+  for (auto &it : this->sessions)
+  {
+    auto res =
+      _data->frame[it.first].StepTo(it.second,
+        std::chrono::duration<double>(_info.simTime).count());
+    if (res.has_value())
+    {
+      it.second = res.value();
+    }
+    else
+    {
+      this->finishedTime = true;
+      return;
+    }
+  }
+
+  // Publish at 2 hz for now. In future make reconfigureable.
+  if (dt.count() > 0.5 && !this->finishedTime)
+  {
+    this->Visualize(_data, _xSamples, _ySamples, _zSamples);
+    this->Publish();
+    lastTick = now;
+  }
+}
+
+/////////////////////////////////////////////////
+void EnvironmentVisualizationTool::Visualize(
+    const std::shared_ptr<components::EnvironmentalData> data,
+    double _xSamples, double _ySamples, double _zSamples)
+{
+
+  for (auto key : data->frame.Keys())
+  {
+    const auto session = this->sessions[key];
+    auto frame = data->frame[key];
+    auto [lower_bound, upper_bound] = frame.Bounds(session);
+    auto step = upper_bound - lower_bound;
+    auto dx = step.X() / _xSamples;
+    auto dy = step.Y() / _ySamples;
+    auto dz = step.Z() / _zSamples;
+    std::size_t idx = 0;
+    for (std::size_t x_steps = 0; x_steps < ceil(_xSamples); x_steps++)
+    {
+      auto x = lower_bound.X() + x_steps * dx;
+      for (std::size_t y_steps = 0; y_steps < ceil(_ySamples); y_steps++)
+      {
+        auto y = lower_bound.Y() + y_steps * dy;
+        for (std::size_t z_steps = 0; z_steps < ceil(_zSamples); z_steps++)
+        {
+          auto z = lower_bound.Z() + z_steps * dz;
+          auto res = frame.LookUp(session, math::Vector3d(x, y, z));
+
+          if (res.has_value())
+          {
+            this->floatFields[key].mutable_data()->Set(idx,
+              static_cast<float>(res.value()));
+          }
+          else
+          {
+            this->floatFields[key].mutable_data()->Set(idx, std::nanf(""));
+          }
+          idx++;
+        }
+      }
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+void EnvironmentVisualizationTool::Publish()
+{
+  pcPub.Publish(this->pcMsg);
+  for(auto &[key, pub] : this->pubs)
+  {
+    pub.Publish(this->floatFields[key]);
+  }
+}
+
+/////////////////////////////////////////////////
+void EnvironmentVisualizationTool::ResizeCloud(
+  const std::shared_ptr<components::EnvironmentalData> _data,
+  const EntityComponentManager& _ecm,
+  unsigned int _xSamples, unsigned int _ySamples, unsigned int _zSamples)
```

Comment:
Would recommend using different variable names here, because `_xSamples _ySamples _zSamples` is an overloaded name pattern - they are `double`s elsewhere in the same file.